activerecord/lib/active_record/type/date.rb

```diff
@@ -9,6 +9,10 @@ def klass
         ::Date
       end
 
+      def type_cast_for_database(value)
```

Comment:
@kamipo I'm not sure this is sufficient to fix the underlying problem.  I tried this same solution last night (as module included onto `ActiveRecord::Type::Date` to monkey patch it), and I am seeing this method called in some cases, such as when persisting records; however, when calling `where` on the model, it is still coercing to a UTC date time string.  Here is an example (rspec) test:

```ruby
describe 'querying dates with times works around day boundaries' do
    let(:model_class) do
      model_class = Class.new(ActiveRecord::Base)
      model_class.table_name = 'date_serialization_test'
      model_class
    end

    let!(:old_record)       { model_class.create!(a_date_field: Date.new(2016, 6, 8)) }
    let!(:boundary_record)  { model_class.create!(a_date_field: Date.new(2016, 6, 9)) }
    let!(:newer_record)     { model_class.create!(a_date_field: Date.new(2016, 6, 10)) }

    before(:context) do
      ActiveRecord::Migration.new.create_table(:date_serialization_test) do |t|
        t.date :a_date_field
      end
    end

    after(:context) do
      ActiveRecord::Migration.new.drop_table(:date_serialization_test)
    end

    it 'coerces TimeWithZone to Date in current locale' do
      date = Time.new(2016, 6, 9, 23, 0, 0, '-05:00').in_time_zone
      byebug
      query = model_class.where('a_date_field >= ?', date)

      expect(query).to include boundary_record, newer_record
      expect(query).not_to include old_record
    end

    it 'coerces Time to Date in current locale' do
      date = Time.new(2016, 6, 9, 23, 0, 0, '-05:00')
      query = model_class.where('a_date_field >= ?', date)

      expect(query).to include boundary_record, newer_record
      expect(query).not_to include old_record
    end

    it 'coerces DateTime to Date in current locale' do
      date = DateTime.new(2016, 6, 9, 23, 0, 0, '-5')
      query = model_class.where('a_date_field >= ?', date)
      binding.pry
      expect(query).to include boundary_record, newer_record
      expect(query).not_to include old_record
    end

    it 'queries still work correctly using a Date object' do
      date = Date.new(2016, 6, 9)
      query = model_class.where('a_date_field >= ?', date)

      expect(query).to include boundary_record, newer_record
      expect(query).not_to include old_record
    end
  end
```

I suspect that because the `where` clause uses an SQL string, it may not be able to determine the column type?  I am finding that my issue may not be quite the same as the case this and the relate rails 5 PR solves.